### PR TITLE
Maven-independent logic is coupled with maven mojos.

### PR DESCRIPTION
**Pull request by piddubnyi**
Soving #38
#Instances.class is extracted to another package and made public.
Reviewer - please advice a right way to avoid ExcessiveImports check fail.

### CODE REVIEW

**Comment by dmarkov** [comment]
@piddubnyi Thanks, someone will review your pull request soon

**Comment by dmarkov** [comment]
@mkordas could you please review this PR

**Comment by mkordas** [comment]
@piddubnyi I'm on it

File: src/main/java/com/jcabi/dynamodb/core/package-info.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * at <a href="http://www.jcabi.com">www.jcabi.com</a>):
+ *
+ * <pre>&lt;depedency&gt;
+ *   &lt;groupId&gt;com.jcabi&lt;/groupId&gt;
+ *   &lt;artifactId&gt;jcabi-dynamodb-maven-plugin&lt;/artifactId&gt;
+ * &lt;/dependency&gt;</pre>
+ *
+ * @author Igor Piddubnyi (igor.piddubnyi@gmail.com)
+ * @version $Id$
+ * @since 0.4
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/package-info.java]: @piddubnyi should be `0.8`

File: src/main/java/com/jcabi/dynamodb/core/package-info.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * DynamoDB core.
+ *
+ * <p>The only dependency you need is (check our latest version available
+ * at <a href="http://www.jcabi.com">www.jcabi.com</a>):
+ *
+ * <pre>&lt;depedency&gt;
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/package-info.java]: @piddubnyi please see http://stackoverflow.com/a/542142/4647853

File: src/main/java/com/jcabi/dynamodb/core/package-info.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * DynamoDB core.
+ *
+ * <p>The only dependency you need is (check our latest version available
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/package-info.java]: @piddubnyi what's the reason for that comment? Looks like marketing information, not technical documentation for developers

**Comment by mkordas** [comment]
@piddubnyi regarding `/src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java[30-343]: A high number of imports can indicate a high degree of coupling within an object. (ExcessiveImports)` - please suppress it and create puzzle to resolve that issue buy splitting class into smaller ones

**Comment by mkordas** [comment]
@piddubnyi is entire #38 issue being solved here? I'm not convinced...

File: src/main/java/com/jcabi/dynamodb/core/package-info.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * DynamoDB core.
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/package-info.java]: @piddubnyi please give more details about what what is responsibility of "core" logic and what will remain outside

**Comment by mkordas** [comment]
@piddubnyi ping

**Comment by piddubnyi** [comment]
@mkordas I fixed package-info, please check if it's enough.

Regarding  ExcessiveImports, can you please introduce a right way to suppress it, I tried 
'''java @checkstyle ExcessiveImports (400 lines)''' on class level, but my local check is still failing.

I'm not familiar with intention of the plugin also, do you think we need to extract some logic from maven mojos also to complete #38 ?

File: src/main/java/com/jcabi/dynamodb/core/package-info.java (hunk 4 above)
Comment (piddubnyi) [inline src/main/java/com/jcabi/dynamodb/core/package-info.java]: My intention was to extract non-maven logic, and I found that it is only Instances class. Maybe name "core" is not a perfect fit, what is your opinion?

Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/package-info.java]: @piddubnyi first of all, please add '@mkordas' at the beginning of your message: http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html#4.-avoid-noise-—-address-your-comments

File: src/main/java/com/jcabi/dynamodb/core/package-info.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * <pre>
+ * {@code
+ * &lt;depedency&gt;
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/package-info.java]: @piddubnyi if you use `{@code` then you don't need `&lt;`

File: src/main/java/com/jcabi/dynamodb/core/package-info.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ *   &lt;groupId&gt;com.jcabi&lt;/groupId&gt;
+ *   &lt;artifactId&gt;jcabi-dynamodb-maven-plugin&lt;/artifactId&gt;
+ * &lt;/dependency&gt;
+ * }
+ * </pre>
+ *
+ * @author Igor Piddubnyi (igor.piddubnyi@gmail.com)
+ * @version $Id$
+ * @since 0.8
+ * @see <a href="http://www.jcabi.com/jcabi-dynamodb-maven-plugin">project website</a>
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/package-info.java]: @piddubnyi I don't think that this link provides more info about this package

**Comment by mkordas** [comment]
@piddubnyi we need to extract logic from Mojos to this package, moving only one class is not enough

**Comment by mkordas** [comment]
@piddubnyi here you have example how to suppress checks: https://github.com/jcabi/jcabi-dynamodb-maven-plugin/blob/d3a840de77da06ba87d339ecfb2a0fe027f515cb/src/main/java/com/jcabi/dynamodb/maven/plugin/Instances.java#L51

File: src/main/java/com/jcabi/dynamodb/core/package-info.java (hunk 4 above)
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/package-info.java]: @piddubnyi name `core` is fine, another name could be `api`. In Javadoc we need to describe main functionality of the code we're going to migrate here.

**Comment by mkordas** [comment]
@piddubnyi do you need any help here?

**Comment by piddubnyi** [comment]
@mkordas Please identify what exact logic should be extracted to the new package.
As you can see I added suppress, but build is still failing, looks like qulice bug.

File: src/main/java/com/jcabi/dynamodb/core/package-info.java (hunk 5 above)
Comment (piddubnyi) [inline src/main/java/com/jcabi/dynamodb/core/package-info.java]: @mkordas But I thought `&lt;` is used to provides maven artifact formatting, e.g
`
<dependency>
  <groupId>org.apache.commons</groupId>
  <artifactId>commons-lang3</artifactId>
</dependency>
`

**Comment by mkordas** [comment]
@piddubnyi it's not the task of code reviewer to identify what needs to be done, I'm just doing review of changes here :)

File: src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java [hunk 7]
```diff
@@ -66,6 +67,7 @@
  * @version $Id$
  * @checkstyle ClassDataAbstractionCoupling (300 lines)
  * @checkstyle MultipleStringLiterals (300 lines)
+ * @checkstyle ExcessiveImports (400 lines)
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java]: @piddubnyi `ExcessiveImports` is PMD rule, checkstyle exclude will not work with it

**Comment by mkordas** [comment]
@piddubnyi I don't think it's a bug in Qulice, we have a lot of excludes in various projects and they work fine

File: src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java (hunk 7 above)
Comment (piddubnyi) [inline src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java]: @mkordas I checked several teamed projectes and have not found example of ыгззкуыштп ExcessiveImports check, can you provide me one that is working with PMD rules?

**Comment by piddubnyi** [comment]
@mkordas First you saying that moving only one class is not enough, than that it's not your task to  identify what needs to be done. Can we agree on something constructive here (Add a puzzle, move additional logic, etc.) ?

**Comment by mkordas** [comment]
@piddubnyi yes, move additional logic. Just take at least one more Mojo file and transfer that logic to core package. Please add puzzle for the rest.

File: src/main/java/com/jcabi/dynamodb/core/package-info.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * DynamoDB core package.
+ * Logic for running local dynamodb directly.
+ *
+ * <pre>
+ * {@code
+ * &lt;depedency&gt;
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/package-info.java]: @piddubnyi replace `&lt;` with `<` in all places

File: src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java (hunk 7 above)
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java]: @piddubnyi just use search option in GitHub: https://github.com/search?utf8=%E2%9C%93&q=ExcessiveImports+user%3Ateamed&type=Code&ref=searchresults

File: src/main/java/com/jcabi/dynamodb/core/Tables.java [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+
+    /**
+     * AWS endpoint.
+     */
+    private final transient String endpoint;
+
+    /**
+     * Port to connect.
+     */
+    private final Integer port;
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/Tables.java]: @piddubnyi should be `transient`

File: src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java [hunk 10]
```diff
@@ -61,11 +42,9 @@
 
 /**
  * Creates DynamoDB tables.
- *
  * @author Carlos Miranda (miranda.cma@gmail.com)
  * @version $Id$
- * @checkstyle ClassDataAbstractionCoupling (300 lines)
- * @checkstyle MultipleStringLiterals (300 lines)
+ * @checkstyle ProtectedMethodInFinalClassCheck (100 lines)
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java]: @piddubnyi why we need `protected` method in `final` class?

File: src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java [hunk 11]
```diff
@@ -111,232 +80,18 @@
...
-                    stream.close();
-                } catch (final IOException ex) {
-                    Logger.error(
-                        this,
-                        "Failed to close stream with message %s",
-                        ex.getMessage()
-                    );
-                }
-            }
+            throw new MojoFailureException(ex.getMessage());
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java]: @piddubnyi we are loosing stacktrace here, I don't think it is acceptable

File: src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java [hunk 12]
```diff
@@ -111,232 +80,18 @@
...
-        JsonObject json = null;
-        try {
-            stream = new FileInputStream(file);
-            json = Json.createReader(stream).readObject();
+            new Tables(
+                this.tables,
+                this.endpoint,
+                this.tcpPort(),
+                this.key,
+                this.secret
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java]: @piddubnyi please put in one line as much as possible

File: src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java [hunk 13]
```diff
@@ -111,232 +80,18 @@
...
+            )
+                .create(instances);
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java]: @piddubnyi can you move indentation back by 4 spaces? Please see http://www.yegor256.com/2014/10/23/paired-brackets-notation.html#fluent

File: src/main/java/com/jcabi/dynamodb/core/Tables.java [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+            "PMD.ModifiedCyclomaticComplexity",
+            "PMD.AvoidInstantiatingObjectsInLoops"
+        }
+    )
+public final class Tables {
+
+    /**
+     * Location of the tables to be created.
+     */
+    private final transient Collection<String> tables;
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/Tables.java]: @piddubnyi rename to `location`

File: src/main/java/com/jcabi/dynamodb/core/Tables.java [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+import javax.json.Json;
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import javax.json.JsonValue;
+
+/**
+ * Handles DynamoDB tables.
+ * @author Igor Piddubnyi (igor.piddubnyi@gmail.com)
+ * @version $Id$
+ * @since 1.0
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/Tables.java]: @piddubnyi should be `0.8`

File: src/main/java/com/jcabi/dynamodb/core/Tables.java [hunk 16]
```diff
@@ -0,0 +1,347 @@
...
+import java.io.InputStream;
+import java.util.Collection;
+import java.util.LinkedList;
+import javax.json.Json;
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import javax.json.JsonValue;
+
+/**
+ * Handles DynamoDB tables.
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/Tables.java]: @piddubnyi this is description of a job. We should rather describe this object if possible

**Comment by mkordas** [comment]
@piddubnyi please see a few comments above

**Comment by mkordas** [comment]
@piddubnyi ping

**Comment by mkordas** [comment]
@piddubnyi do you need some help with this pull request?

**Comment by piddubnyi** [comment]
@mkordas Done.

**Comment by mkordas** [comment]
@piddubnyi I'm on it again

File: src/main/java/com/jcabi/dynamodb/core/package-info.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * DynamoDB core package. Logic for running local dynamodb directly.
+ *
+ * <pre>&lt;depedency&gt;
+ *   &lt;groupId&gt;com.jcabi&lt;/groupId&gt;
+ *   &lt;artifactId&gt;jcabi-dynamodb-maven-plugin&lt;/artifactId&gt;
+ * &lt;/dependency&gt;</pre>
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/package-info.java]: @piddubnyi please remove that dependency information, it is not needed here

File: src/main/java/com/jcabi/dynamodb/core/Tables.java [hunk 18]
```diff
@@ -0,0 +1,347 @@
...
+import javax.json.Json;
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import javax.json.JsonValue;
+
+/**
+ * Handles DynamoDB locations.
+ * @author Igor Piddubnyi (igor.piddubnyi@gmail.com)
+ * @version $Id$
+ * @since 1.0
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/Tables.java]: @piddubnyi should be `0.8`

File: src/main/java/com/jcabi/dynamodb/core/Tables.java [hunk 19]
```diff
@@ -0,0 +1,347 @@
...
+import java.util.Collection;
+import java.util.LinkedList;
+import javax.json.Json;
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import javax.json.JsonValue;
+
+/**
+ * Handles DynamoDB locations.
+ * @author Igor Piddubnyi (igor.piddubnyi@gmail.com)
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/core/Tables.java]: @piddubnyi please also preserve original author

File: src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java [hunk 20]
```diff
@@ -112,231 +80,14 @@
...
-                    stream.close();
-                } catch (final IOException ex) {
-                    Logger.error(
-                        this,
-                        "Failed to close stream with message %s",
-                        ex.getMessage()
-                    );
-                }
-            }
+            throw new MojoFailureException(ex.getMessage(), ex);
```
Comment (mkordas) [inline src/main/java/com/jcabi/dynamodb/maven/plugin/CreateTablesMojo.java]: @piddubnyi please read http://www.yegor256.com/2015/12/01/rethrow-exceptions.html and add proper context

**Comment by mkordas** [comment]
@piddubnyi couple of additional comments above, generally it looks nice now

**Comment by piddubnyi** [comment]
@mkordas please check.

**Comment by mkordas** [comment]
@piddubnyi looks good

**Comment by mkordas** [comment]
@rultor merge

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by mkordas** [comment]
@rultor deploy

**Comment by dmarkov** [comment]
@mkordas Thanks a lot, I just topped your account for **56 mins**, transaction ID `56b845b7d4cb5536f3000225`, total time was 211 hours and 10 mins... extra minutes for review comments (c=41)... +56 added to your rating, at the moment it is: [+5526](http://www.netbout.com/b/36080?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy

**Comment by amihaiemil** [comment]
@mkordas A puzzle was left in ``package-info.java``, saying that logic should be taken out of the mojos, in a different package so it can be used separately... it's ticket #52 . 
I look around and I see that, for now, this was already done in this PR.
 
This puzzle is more like a continuous todo for developers - future logic development should be separated from the mojos. I'm not sure if this concept exists in XDSD, but my point is that currently the puzzle is invalid, since I see no more logic that can be separated from the mojos... 

Have you stumbled on this kind of sitiuation before? What should I do - create a PR in which I simply remove the puzzle?

**Comment by mkordas** [comment]
@amihaiemil you can make a PR with removing puzzle with a proof that there is nothing else to do. Another option is to create test confirming that indeed logic can be used on its own now.

**Pull request merged**